File: Day_5/Print_Queue.go
package main

import (
	"fmt"
	"os"
	"strings"
	"strconv"
)

func check(e error){
	if e != nil{
		panic(e)
	}
}

var print = fmt.Println

func contains(slice []int, element int) bool {
    for _, v := range slice {
        if v == element {
            return true
        }
    }
    return false
}

func check_pages(line []int, dic map[int][]int) bool{
	for i := 0; i < len(line)-1; i++{
		for j := i+1; j<len(line); j++{
			if _, ok := dic[line[i]]; ok{
				if contains(dic[line[i]], line[j]){
					continue
				}else{
					if contains(dic[line[j]], line[i]){
						return false
					}
				}
			}else{
				if _, ok := dic[line[j]]; ok{
					if contains(dic[line[j]], line[i]){
						return false
					}
				}
			}
			
		}
	}
	return true
}

func find_wrong(wrong []int, dic map[int][]int) []int{
	for i := 0; i < len(wrong)-1; i++{
		for j := i+1; j < len(wrong); j++{
			if _, ok := dic[wrong[i]]; ok{
				if contains(dic[wrong[i]], wrong[j]){
					continue
				}else{
					if contains(dic[wrong[j]], wrong[i]){
						wrong[i], wrong[j] = wrong[j], wrong[i]
						if check_pages(wrong, dic){
							return wrong
						}else{
							return find_wrong(wrong, dic)
						}
					}
				}
			}else{
				if _, ok := dic[wrong[j]]; ok{
					if contains(dic[wrong[j]], wrong[i]){
						wrong[i], wrong[j] = wrong[j], wrong[i]
						if check_pages(wrong, dic){
							return wrong
						}else{
							return find_wrong(wrong, dic)
						}
					}
				}
			}
		}
	}
	return wrong
}

func main(){
	dat, err := os.ReadFile("D5.txt")
	check(err)
	data := strings.Split(strings.ReplaceAll(string(dat), "\r\n", "\n"), "\n\n")
	page_com := strings.Split(data[0], "\n")
	pages := make([][]int, 0)
	for _, val := range strings.Split(data[1], "\n"){
		temp_int := make([]int, 0)
		for _, val := range strings.Split(val, ","){
			a := 0
			fmt.Sscanf(val, "%d", &a)
			temp_int = append(temp_int, a)
		}
		pages = append(pages, temp_int)
	}
	dic := make(map[int][]int)
	for _, val := range page_com{
		key, _ := strconv.Atoi(val[:2])
		content, _ := strconv.Atoi(val[3:])
		dic[key] = append(dic[key], content)

	}
	total := 0
	wrong := make([][]int, 0)
	for _, val := range pages{
		if check_pages(val, dic){
			total+= val[len(val)/2]
		}else{
			wrong = append(wrong, val)
		}
	}
	print(total)

	total = 0
	for _, val := range wrong{
		temp := find_wrong(val, dic)
		total+= temp[len(temp)/2]
	}
	print(total)
}

File: Day_5/Print_Queue.py
def check_pages(line, dic) -> bool:
    for i in range(len(line)-1):
        for j in range(i+1, len(line)):
            if line[i] in dic.keys():
                if line[j] in dic[line[i]]:
                    continue
                else:
                    if line[i] in dic[line[j]]:
                        return False
            else:
                if line[j] in dic.keys():
                    if line[i] in dic[line[j]]:
                        return False

    return True

# Part 1

with open("D5.txt", 'r') as file:
    data = file.read().split("\n\n")
page_com = data[0].split("\n")
pages = [list(map(int, i.split(','))) for i in data[1].split("\n")]
dic = {} 

for i in range(len(page_com)):
    dic[int(page_com[i][:2])] = []
for i in range(len(page_com)):
    dic[int(page_com[i][:2])].append(int(page_com[i][3:]))

total = 0
wrong = []
for i in pages:
    if check_pages(i, dic):
        # print(i)
        total += i[len(i)//2]
    else:
        wrong.append(i)
print(total)
# 4637

# Part 2
def find_wrong(wrong, dic):
     for i in range(len(wrong)-1):
        for j in range(i+1, len(wrong)):
            if wrong[i] in dic.keys():
                if wrong[j] in dic[wrong[i]]:
                    continue
                else:
                    if wrong[i] in dic[wrong[j]]:
                        wrong[i], wrong[j] = wrong[j], wrong[i]
                        if check_pages(wrong, dic):
                            return wrong
                        else:
                            return find_wrong(wrong, dic)
            else:
                if wrong[j] in dic.keys():
                    if wrong[i] in dic[wrong[j]]:
                        wrong[i], wrong[j] = wrong[j], wrong[i]
                        if check_pages(wrong, dic):
                            return wrong
                        else:
                            return find_wrong(wrong, dic)

total = 0
for i in wrong:
    temp =  find_wrong(i, dic)
    print(temp)
    total += temp[len(temp)//2]
print(total)
# 6370
